xbmc/threads/Thread.cpp

```diff
@@ -79,61 +71,98 @@ void CThread::Create(bool bAutoDelete, unsigned stacksize)
   m_bAutoDelete = bAutoDelete;
   m_bStop = false;
   m_StopEvent.Reset();
-  m_TermEvent.Reset();
   m_StartEvent.Reset();
 
-  SpawnThread(stacksize);
-}
-
-bool CThread::IsRunning() const
-{
-  return m_ThreadId ? true : false;
-}
-
-THREADFUNC CThread::staticThread(void* data)
-{
-  CThread* pThread = static_cast<CThread*>(data);
-  std::string name;
-  ThreadIdentifier id;
-  bool autodelete;
-
-  if (!pThread) {
-    CLog::Log(LOGERROR,"%s, sanity failed. thread is NULL.",__FUNCTION__);
-    return 1;
-  }
-
-  name = pThread->m_ThreadName;
-  id = GetDisplayThreadId(pThread->m_ThreadId);
-  autodelete = pThread->m_bAutoDelete;
+  // lock?
+  //CSingleLock l(m_CriticalSection);
 
-  pThread->SetThreadInfo();
+  std::promise<bool> prom;
+  m_future = prom.get_future();
 
-  CLog::Log(LOGDEBUG,"Thread %s start, auto delete: %s", name.c_str(), (autodelete ? "true" : "false"));
-
-  currentThread = pThread;
-  pThread->m_StartEvent.Set();
-
-  pThread->Action();
-
-  // lock during termination
-  CSingleLock lock(pThread->m_CriticalSection);
-
-  pThread->m_ThreadId = 0;
-  pThread->m_TermEvent.Set();
-  pThread->TermHandler();
-
-  lock.Leave();
-
-  if (autodelete)
   {
-    CLog::Log(LOGDEBUG,"Thread %s %" PRIu64" terminating (autodelete)", name.c_str(), (uint64_t)id);
-    delete pThread;
-    pThread = NULL;
-  }
-  else
-    CLog::Log(LOGDEBUG,"Thread %s %" PRIu64" terminating", name.c_str(), (uint64_t)id);
+    // The std::thread internals must be set prior to the lambda doing
+    //   any work. This will cause the lambda to wait until m_thread
+    //   is fully initialized. Interestingly, using a std::atomic doesn't
+    //   have the appropriate memory barrier behavior to accomplish the
+    //   same thing so a full system mutex needs to be used.
+    CSingleLock blockLambdaTillDone(m_CriticalSection);
+    m_thread = new std::thread([](CThread* pThread, std::promise<bool> promise)
+    {
+      try
+      {
+
+        {
+          // Wait for the pThread->m_thread internals to be set. Otherwise we could
+          // get to a place where we're reading, say, the thread id inside this
+          // lambda's call stack prior to the thread that kicked off this lambda
+          // having it set. Once this lock is released, the CThread::Create function
+          // that kicked this off is done so everything should be set.
+        CSingleLock waitForThreadInternalsToBeSet(pThread->m_CriticalSection);
```

Comment:
wrong indentation till line 157

Reply:
And the color of the bikeshed shall be ... BLUE!